src/elements/image.rs

```diff
@@ -15,6 +15,7 @@ impl Image {
         Self::default()
     }
 
+    /// The `Data` should, and must outlive the `Clay` instance !!!
     pub fn data<Data>(&mut self, data: &mut Data) -> &mut Self {
```

Comment:
We can enforce this with proper lifetimes.

Reply:
I personaly have no idea on how to enforce it, but yeah, having a safer approach is not only a good idea, but a must

Reply:
You do something like this (untested)

```Rust
struct Clay<'a> {
   ...
}

impl Clay {
   pub fn data<Data>(&mut self, data: &'a Data) {}
}
```
This will ensure that `Data` has to outlive Clay due to the lifetime bounds.

Reply:
Oh, nevermind. This won't work as `Image` builder isn't part of Clay object. I will have to think a bit more on it.

Reply:
I might have an idea, we could use the clay object with something like `Clay::store` that ensures the right lifetime, and it returns a wrapper struct to a `*mut c_void` that can then be passed as data to images and custom

Reply:
That isn't great because the user may have several images that gets used with multiple `Image::new()` calls. What would be possible to do instead is to have `clay.image_new().data(...)` That we we could propagate the lifetime from the Clay object to the builder for `Image`. The issue here is that the API then becomes different from the other builders.

Reply:
I think the `store` doesn't disallow for using several images with multiple `Image::new()` calls

Reply:
Well it doesn't really disallow, but it's a bit weird API imo. Something perhaps better would be this

```Rust

let data_id  = clay.data(...);
clay.with([
  Image::new(...).data(data_id);
  ...
```
Now the user is forced to call `clay.data(...)` to get a id back that can then be used with the builder.

Reply:
The `clay.store` api is pretty much the same, just instead of passing an id it passes a struct wrapping the pointer to the data

Reply:
Oh, I see now this is likely what you meant already :laughing:

Reply:
I think a better name is `get_id` or `data` instead of `store` tho.

Reply:
Yeah `Clay::data` seems better